File: src/justForReal/ThreeSum.java
package justForReal;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class ThreeSum {
	public List<List<Integer>> threeSum(int[] nums) {
		Arrays.sort(nums);
		List<List<Integer>> result = new LinkedList<>();
		for (int i = 0; i < nums.length - 2; i++) {
			if (i == 0 || (i > 0 && nums[i] != nums[i - 1])) {
				int lo = i + 1;
				int hi = nums.length - 1;
				while (lo < hi) {
					if (nums[i] + nums[lo] + nums[hi] == 0) {
						result.add(Arrays.asList(nums[i], nums[lo], nums[hi]));
						while (lo < hi && nums[lo] == nums[lo + 1])
							lo++;
						while (lo < hi && nums[hi] == nums[hi - 1])
							hi--;
						lo++;
						hi--;
					} else if (nums[i] + nums[lo] + nums[hi] < 0) {
						lo++;
					} else {
						hi--;
					}

				}
			}
		}
		return result;
	}
}

File: src/justForReal/threeSum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for i in range(len(nums)-2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            a,b = i+1,len(nums)-1
            temp = -nums[i]
            while(b>a):
                if temp==(nums[b]+nums[a]):
                    # t_res = [nums[i],nums[a],nums[b]]
                    # if t_res not in res:
                    #     res.append(t_res)
                    # a+=1
                    while a < len(nums) - 1 and nums[a] == nums[a + 1]:
                        a += 1
                    while b > 0 and nums[b] == nums[b - 1]:
                        b -= 1
                    res.append([nums[i], nums[a], nums[b]])
                    a += 1
                elif temp>(nums[b]+nums[a]):
                    a+=1
                else:
                    b-=1
        return res